tailnet/multiagent.go

```diff
@@ -0,0 +1,154 @@
+package tailnet
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+)
+
+type MultiAgentConn interface {
+	UpdateSelf(node *Node) error
+	SubscribeAgent(agentID uuid.UUID) error
+	UnsubscribeAgent(agentID uuid.UUID) error
+	NextUpdate(ctx context.Context) ([]*Node, bool)
+	AgentIsLegacy(agentID uuid.UUID) bool
+	Close() error
+	IsClosed() bool
+}
+
+type MultiAgent struct {
+	mu sync.RWMutex
+
+	closed bool
+
+	ID     uuid.UUID
+	Logger slog.Logger
+
+	AgentIsLegacyFunc func(agentID uuid.UUID) bool
+	OnSubscribe       func(enq Queue, agent uuid.UUID) error
+	OnUnsubscribe     func(enq Queue, agent uuid.UUID) error
+	OnNodeUpdate      func(id uuid.UUID, node *Node) error
+	OnRemove          func(id uuid.UUID)
+
+	updates   chan []*Node
+	closeOnce sync.Once
+	start     int64
+	lastWrite int64
+	// Client nodes normally generate a unique id for each connection so
+	// overwrites are really not an issue, but is provided for compatibility.
+	overwrites int64
+}
+
+func (m *MultiAgent) Init() *MultiAgent {
+	m.updates = make(chan []*Node, 128)
+	m.start = time.Now().Unix()
+	return m
+}
+
+func (m *MultiAgent) UniqueID() uuid.UUID {
+	return m.ID
+}
+
+func (m *MultiAgent) AgentIsLegacy(agentID uuid.UUID) bool {
+	return m.AgentIsLegacyFunc(agentID)
+}
+
+var ErrMultiAgentClosed = xerrors.New("multiagent is closed")
+
+func (m *MultiAgent) UpdateSelf(node *Node) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if m.closed {
+		return ErrMultiAgentClosed
+	}
+
+	return m.OnNodeUpdate(m.ID, node)
+}
+
+func (m *MultiAgent) SubscribeAgent(agentID uuid.UUID) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if m.closed {
+		return ErrMultiAgentClosed
+	}
+
+	return m.OnSubscribe(m, agentID)
```

Comment:
This can cause a deadlock if something attempts to `Close()` the multi-agent.

> If a goroutine holds a RWMutex for reading and another goroutine might call Lock, no goroutine should expect to be able to acquire a read lock until the initial read lock is released. In particular, this prohibits recursive read locking. This is to ensure that the lock eventually becomes available; a blocked Lock call excludes new readers from acquiring the lock.
>
> https://pkg.go.dev/sync#RWMutex

`m.OnSubscribe()` can result in a call back to `Enqueue()` if the node for the agent is known, resulting in reentrant call to `RLock()`.